packages/flutter/lib/src/widgets/localizations.dart

```diff
@@ -0,0 +1,474 @@
+  static Locale localeOf(BuildContext context) {
+    assert(context != null);
+    final _LocalizationsScope scope = context.inheritFromWidgetOfExactType(_LocalizationsScope);
+    return scope.localizedResourcesState.locale;
+  }
+
+  /// Returns the 'type' localized resources for the widget tree that
+  /// corresponds to [BuildContext] `context`.
+  ///
+  /// This method is typically used by a static factory method on the 'type'
+  /// class. For example Flutter's MaterialLocalizations class looks up Material
+  /// resources with a method defined like this:
+  ///
+  /// ```dart
+  /// static MaterialLocalizations of(BuildContext context) {
+  ///    return Localizations.of<MaterialLocalizations>(context, MaterialLocalizations);
+  /// }
+  /// ```
+  static T of<T>(BuildContext context, Type type) {
+    assert(context != null);
+    assert(type != null);
+    final _LocalizationsScope scope = context.inheritFromWidgetOfExactType(_LocalizationsScope);
+    return scope.localizedResourcesState.resourcesFor<T>(type);
+  }
+
+  @override
+  _LocalizationsState createState() => new _LocalizationsState();
+}
+
+class _LocalizationsState extends State<Localizations> {
+  final GlobalKey _localizedResourcesScopeKey = new GlobalKey();
+
+  Locale get locale => _locale;
+  Locale _locale;
+
+  @override
+  void initState() {
+    super.initState();
+    if (widget.delegate == null)
+      _locale = widget.locale;
+    else
+      load(widget.locale);
+  }
+
+  @override
+  void didUpdateWidget(Localizations old) {
+    super.didUpdateWidget(old);
+    if (widget.delegate != old.delegate && widget.delegate != null)
+      load(widget.locale);
+  }
+
+  void load(Locale locale) {
+    final LocalizationsDelegate delegate = widget.delegate;
+    if (delegate == null) {
+      _locale = locale;
+      return;
+    }
+
+    Iterable<Type> types;
+    final Future<Iterable<Type>> futureTypes = delegate.load(locale)
+      .then<Iterable<Type>>((Iterable<Type> value) {
+        return types = value;
+      });
+
+    if (types != null) {
+      // The delegate loaded its resources synchronously, do not rebuild
+      _locale = locale;
+    } else {
+      // Don't rebuild the dependent widgets until the resources for the new locale
+      // have finished loading. Until then the old locale will continue to be used.
+      futureTypes.then<Iterable<Type>>((Iterable<Type> value) {
+        setState(() {
+          _locale = locale;
+        });
+        final InheritedElement scopeElement = _localizedResourcesScopeKey.currentContext;
+        scopeElement?.dispatchDidChangeDependencies();
+      });
+    }
+  }
+
+  T resourcesFor<T>(Type type) => widget.delegate?.resourcesFor<T>(_locale, type);
+
+  @override
+  Widget build(BuildContext context) {
+    return new _LocalizationsScope(
+      key: _localizedResourcesScopeKey,
+      locale: widget.locale,
+      localizedResourcesState: this,
+      child: _locale != null ? widget.child : new Container(),
```

Comment:
If we're going to have a splash screen facility then it might be better to just defer the "first frame" signal that takes the splash screen down, until the initial async localizations have finished loading.

I've updated https://github.com/flutter/flutter/issues/1865